### DISCUSSION TITLE
Simpler syntax for iterating over a fixed set of values in `foreach`

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#7067] [Language Ideas] [answered]
## Summary

Introduce the ability to remove the need for `new[]` or `new int[]` behind an array within the expression for `in` in a `foreach` loop.

## Motivation

I have written some tests and I want to iterate through some fixed hand-picked values for parameters. Here is the example:
```csharp
foreach (var neighborCount in new[] { 12, 14, 29, 58, 124, 178, 341, 589 })
{
   // use neighborCount
}
```

I would like to be able to use something closer to the variable initializer syntax, omitting the need for `new[]` and implying the array type:
```csharp
foreach (var neighborCount in { 12, 14, 29, 58, 124, 178, 341, 589 })
{
   // use neighborCount
}
```

> **NOTE:** Implicit array initializers do not work on `var` assignments. Introducing this inference like in the above example would feel more powerful than its natural and existing counterpart in variable and field initializers.
> However, to align with the currently available inference, this feature would be available by specifying the type of the `foreach` iterator variable. This would then look like this:
```csharp
foreach (int neighborCount in { 12, 14, 29, 58, 124, 178, 341, 589 })
{
   // use neighborCount
}
```

**Comment by HaloFour** [comment, marked as answer]
With #5354 you should be able to do the following:

```cs
foreach (int neighborCount in [ 12, 14, 29, 58, 124, 178, 341, 589 ])
{
   // use neighborCount
}
```

**Reply by jnm2** [reply]
One thing I'll be enjoying is `foreach (var someCond in [false, true])`.

**Reply by Enderlook** [reply]
It would be great if that use case would leave as an implementation detail how it's actually mapped, so the compiler doesn't necessarily must allocate arrays or lists, but instead can allocate spans or even unroll the loop in some future.
For example, something like:
```cs
foreach (int neighborCount in [ 12, 14, 29, 58, someVariable, 178, 341, 589 ])
{
   // ..
}
```
Could be mapped into:
```cs
private static ReadOnlySpan<int> __const1 => new[] { 12, 14, 29, 58 }; // This doesn't allocate as the compiler embeds the content in the assembly.
private static ReadOnlySpan<int> __const2 => new[] { 178, 341, 589 };

foreach (int neighborCount in __const1)
{
   // Loop Body
}

int neighborCount = someVariable;
// Loop Body

foreach (int neighborCount in __const2)
{
   // Loop Body
}
```
Of course, in the example I provided, the method could use `stackalloc` and  a single loop:
```cs
foreach (int neighborCount in stackalloc int[] { 12, 14, 29, 58, someVariable, 178, 341, 589 })
{
   // Loop Body
}
```
But for larger enumerations, that would produce `StackOverflowException`.

Or maybe this is just too much for a rare pattern haha.

**Reply by CyrusNajmabadi** [reply]
> It would be great if that use case would leave as an implementation detail how it's actually mapped, so the compiler doesn't necessarily must allocate arrays or lists

That's exactly what the specification says.
